src/module/ovs/ovs.rs

```diff
@@ -29,31 +57,294 @@ impl Collector for OvsCollector {
+        )?)?)?;
+        Ok(())
+    }
+
+    /// Add upcall hooks.
+    fn add_upcall_hooks(&self, probes: &mut ProbeManager) -> Result<()> {
+        let inflight_upcalls_map = self
+            .inflight_upcalls_map
+            .as_ref()
+            .ok_or_else(|| anyhow!("Inflight upcalls map not created"))?
+            .fd();
+
+        // Upcall probe.
+        let mut kernel_upcall_tp_hook = Hook::from(kernel_upcall_tp::DATA);
+        kernel_upcall_tp_hook.reuse_map("inflight_upcalls", inflight_upcalls_map)?;
+        probes.register_hook_to(
+            kernel_upcall_tp_hook,
+            Probe::raw_tracepoint(Symbol::from_name("openvswitch:ovs_dp_upcall")?)?,
+        )?;
+
+        // Upcall return probe.
+        let mut kernel_upcall_ret_hook = Hook::from(kernel_upcall_ret::DATA);
+        kernel_upcall_ret_hook.reuse_map("inflight_upcalls", inflight_upcalls_map)?;
+        probes.register_hook_to(
+            kernel_upcall_ret_hook,
+            Probe::kretprobe(Symbol::from_name("ovs_dp_upcall")?)?,
+        )?;
+
+        if self.track {
+            // Upcall enqueue.
+            let mut kernel_enqueue_hook = Hook::from(kernel_enqueue::DATA);
+            kernel_enqueue_hook.reuse_map("inflight_upcalls", inflight_upcalls_map)?;
+            probes.register_hook_to(
+                kernel_enqueue_hook,
+                Probe::kretprobe(Symbol::from_name("queue_userspace_packet")?)?,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    /// Add exec hooks.
+    fn add_exec_hooks(&mut self, probes: &mut ProbeManager) -> Result<()> {
+        let mut exec_action_hook = Hook::from(kernel_exec_tp::DATA);
+
+        if self.track {
+            let inflight_map = Self::create_inflight_exec_cmd_map()?;
+
+            exec_action_hook.reuse_map("inflight_exec_cmd", inflight_map.fd())?;
+
+            let mut exec_cmd_hook = Hook::from(kernel_exec_cmd::DATA);
+            exec_cmd_hook.reuse_map("inflight_exec_cmd", inflight_map.fd())?;
+            probes.register_hook_to(
+                exec_cmd_hook,
+                Probe::kprobe(Symbol::from_name("ovs_packet_cmd_execute")?)?,
+            )?;
+
+            let mut exec_cmd_ret_hook = Hook::from(kernel_exec_cmd_ret::DATA);
+            exec_cmd_ret_hook.reuse_map("inflight_exec_cmd", inflight_map.fd())?;
+            probes.register_hook_to(
+                exec_cmd_ret_hook,
+                Probe::kretprobe(Symbol::from_name("ovs_packet_cmd_execute")?)?,
+            )?;
+
+            self.inflight_exec_cmd_map = Some(inflight_map);
+        }
+
+        // Action execute probe.
+        probes.register_hook_to(
+            exec_action_hook,
+            Probe::raw_tracepoint(Symbol::from_name("openvswitch:ovs_do_execute_action")?)?,
+        )?;
+
+        Ok(())
+    }
+
+    /// Add USDT hooks.
+    fn add_usdt_hooks(&mut self, probes: &mut ProbeManager) -> Result<()> {
+        let ovs = Process::from_cmd("ovs-vswitchd")?;
```

Comment:
It'll only fail if `--ovs-track` is given, right? Isn't that a clear statement from the user that they want OVS userspace to be tracked?

Reply:
> Ah right, I overlooked this was under this conditional. `Process::from_cmd` won't work then, but we could have a way to use the same initial logic of that function to only check if the process is there an running?
> 

Not sure I'm following. You're suggesting we don't fail (just warn or something like that) even if `--ovs-track` is given and OVS is not running?

> An other answer is to have a system "capabilities detection" in the core and use that to dynamically construct the list of modules. Which IMHO is a fine solution and of course not have to be part of this PR.

Reply:
> > > Ah right, I overlooked this was under this conditional. `Process::from_cmd` won't work then, but we could have a way to use the same initial logic of that function to only check if the process is there an running?
> > 
> > 
> > Not sure I'm following. You're suggesting we don't fail (just warn or something like that) even if `--ovs-track` is given and OVS is not running?
> 
> No, I'm suggesting to detect early if OVS is running and do not use the OVS module if not, but not fail the tool.
> 
> The reasoning is not selecting modules explicitly can't work in all cases and I think we shouldn't fail in this case but not use the "problematic" modules while still issuing a warning (or info! would probably be more suited) so the user don't have unmatched expectations.
> 

Ok. For this specific case I don't think detecting if OVS is running is what we want. Without `--ovs-track` we're capable of inspecting the kernel datapath for which we don't care if OVS is running (i.e: accessible in our pid-namespace with a single pid).

So what we want to check is if the `openvswitch` module is present.

> This can be generalized later to have a "system inspector" allowing for checking if a process is running, what version is used, etc. for dynamic actions in modules.

Having a system inspector capable of determining the version of arbitrary userspace applications seems too complicated and I think we might be better off doing it in the module that knows how to deal with the specific application.
What the generic inspector could give us is the kernel version and the list of loaded modules. That could be used to disable ovs module (well ... until we support DPDK datapath :-D)

Reply:
Created #100 to track this.